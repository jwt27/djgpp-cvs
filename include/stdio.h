/* Copyright (C) 2014 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2013 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2003 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2002 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2001 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2000 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1998 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1995 DJ Delorie, see COPYING.DJ for details */

#ifndef __dj_include_stdio_h_
#define __dj_include_stdio_h_

#ifdef __cplusplus
extern "C" {
#endif

#ifndef __dj_ENFORCE_ANSI_FREESTANDING

#include <sys/version.h>
#include <sys/djtypes.h>
  
#define _IOFBF    	00001  /*  File is full buffered.  */
#define _IONBF    	00002  /*  File is not buffered.  */
#define _IOLBF    	00004  /*  File is line buffered.  */

/* Some programs think they know better... */
#undef NULL
#if (__GNUC__ >= 4) && defined(__cplusplus)
#  define NULL          __null
#elif defined(__cplusplus)
#  define NULL          0
#else
#  define NULL          ((void*)0)
#endif

#define BUFSIZ		16384
#define EOF		(-1)
#define FILENAME_MAX	260
#define FOPEN_MAX	20
#define L_tmpnam	260
#define TMP_MAX		999999

#define SEEK_SET	0
#define SEEK_CUR	1
#define SEEK_END	2

#ifndef _VA_LIST
__DJ_va_list
#define _VA_LIST
#endif

#ifndef _SIZE_T
__DJ_size_t
#define _SIZE_T
#endif

#ifndef _SSIZE_T
__DJ_ssize_t
#define _SSIZE_T
#endif

/* Note that the definitions of these fields are NOT guaranteed!  They
   may change with any release without notice!  The fact that they
   are here at all is to comply with ANSI specifictions. */
   
typedef struct {
  ssize_t  _cnt;
  char    *_ptr;
  char    *_base;
  size_t   _bufsiz;
  int      _flag;
  int      _file;
  char    *_name_to_remove;
  size_t   _fillsize;
} FILE;

typedef unsigned long		fpos_t;

extern FILE __dj_stdin, __dj_stdout, __dj_stderr;
#define stdin	(&__dj_stdin)
#define stdout	(&__dj_stdout)
#define stderr	(&__dj_stderr)

void	clearerr(FILE *_stream);
int	fclose(FILE *_stream);
int	feof(FILE *_stream);
int	ferror(FILE *_stream);
int	fflush(FILE *_stream);
int	fgetc(FILE *_stream);
int	fgetpos(FILE *_stream, fpos_t *_pos);
char *	fgets(char *_s, int _n, FILE *_stream);
FILE *	fopen(const char *_filename, const char *_mode);
int	fprintf(FILE *_stream, const char *_format, ...);
int	fputc(int _c, FILE *_stream);
int	fputs(const char *_s, FILE *_stream);
size_t	fread(void *_ptr, size_t _size, size_t _nelem, FILE *_stream);
FILE *	freopen(const char *_filename, const char *_mode, FILE *_stream);
int	fscanf(FILE *_stream, const char *_format, ...);
int	fseek(FILE *_stream, long _offset, int _mode);
int	fsetpos(FILE *_stream, const fpos_t *_pos);
long	ftell(FILE *_stream);
size_t	fwrite(const void *_ptr, size_t _size, size_t _nelem, FILE *_stream);
int	getc(FILE *_stream);
int	getchar(void);
char *	gets(char *_s);
void	perror(const char *_s);
int	printf(const char *_format, ...);
int	putc(int _c, FILE *_stream);
int	putchar(int _c);
int	puts(const char *_s);
int	remove(const char *_filename);
int	rename(const char *_old, const char *_new);
void	rewind(FILE *_stream);
int	scanf(const char *_format, ...);
void	setbuf(FILE *_stream, char *_buf);
int	setvbuf(FILE *_stream, char *_buf, int _mode, size_t _size);
int	sprintf(char *_s, const char *_format, ...);
int	sscanf(const char *_s, const char *_format, ...);
FILE *	tmpfile(void);
char *	tmpnam(char *_s);
int	ungetc(int _c, FILE *_stream);
int	vfprintf(FILE *_stream, const char *_format, va_list _ap);
int	vprintf(const char *_format, va_list _ap);
int	vsprintf(char *_s, const char *_format, va_list _ap);

#if (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L) \
  || !defined(__STRICT_ANSI__) || defined(__cplusplus)

int	snprintf(char *str, size_t n, const char *fmt, ...);
int	vfscanf(FILE *_stream, const char *_format, va_list _ap);
int	vscanf(const char *_format, va_list _ap);
int	vsnprintf(char *str, size_t n, const char *fmt, va_list ap);
int	vsscanf(const char *_s, const char *_format, va_list _ap);

#endif /* (__STDC_VERSION__ >= 199901L) || !__STRICT_ANSI__ */

#ifndef __STRICT_ANSI__

#define L_ctermid 20
#define L_cusrid
/* #define STREAM_MAX	20 - DOS can change this */

int	dprintf(int _fd, const char *_format, ...) __attribute__ ((__format__ (__printf__, 2, 3)));
int	fileno(FILE *_stream);
FILE *	fdopen(int _fildes, const char *_type);
int	mkstemp(char *_template);
int	pclose(FILE *_pf);
FILE *	popen(const char *_command, const char *_mode);
char *	tempnam(const char *_dir, const char *_prefix);
int	vdprintf(int _fd, const char *_format, va_list _ap) __attribute__ ((__format__ (__printf__, 2, 0)));

#ifndef _POSIX_SOURCE

extern FILE __dj_stdprn, __dj_stdaux;
#define stdprn	(&__dj_stdprn)
#define stdaux	(&__dj_stdaux)

#define P_tmpdir "c:/"

void	_djstat_describe_lossage(FILE *_to_where);
int	_doprnt(const char *_fmt, va_list _args, FILE *_f);
int	_doscan(FILE *_f, const char *_fmt, va_list _args);
int	_doscan_low(FILE *, int (*)(FILE *_get), int (*_unget)(int, FILE *), const char *_fmt, va_list _args);
int	fpurge(FILE *_f);
int	getw(FILE *_f);
char *	mktemp(char *_template);
int	putw(int _v, FILE *_f);
void	setbuffer(FILE *_f, void *_buf, int _size);
void	setlinebuf(FILE *_f);
int	_rename(const char *_old, const char *_new);	/* Simple (no directory) */
int	asprintf(char **_sp, const char *_format, ...) __attribute__((format (__printf__, 2, 3)));
char *	asnprintf(char *_s, size_t *_np, const char *_format, ...) __attribute__((format (__printf__, 3, 4)));
int	vasprintf(char **_sp, const char *_format, va_list _ap) __attribute__((format (__printf__, 2, 0)));
char *	vasnprintf(char *_s, size_t *_np, const char *_format, va_list _ap) __attribute__((format (__printf__, 3, 0)));

#ifndef _OFF_T
__DJ_off_t
#define _OFF_T
#endif
#ifndef _OFF64_T
__DJ_off64_t
#define _OFF64_T
#endif
int	fseeko(FILE *_stream, off_t _offset, int _mode);
off_t	ftello(FILE *_stream);
int	fseeko64(FILE *_stream, off64_t _offset, int _mode);
off64_t	ftello64(FILE *_stream);

#endif /* !_POSIX_SOURCE */
#endif /* !__STRICT_ANSI__ */
#endif /* !__dj_ENFORCE_ANSI_FREESTANDING */

#ifndef __dj_ENFORCE_FUNCTION_CALLS
#endif /* !__dj_ENFORCE_FUNCTION_CALLS */

#ifdef __cplusplus
}
#endif

#endif /* !__dj_include_stdio_h_ */
