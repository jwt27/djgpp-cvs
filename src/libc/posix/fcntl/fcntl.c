/* Copyright (C) 2015 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2010 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2003 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2002 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2001 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1999 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1996 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1995 DJ Delorie, see COPYING.DJ for details */
#include <libc/stubs.h>
#include <dpmi.h>
#include <errno.h>
#include <stdarg.h>
#include <unistd.h>
#include <fcntl.h>
#include <io.h>
#include <sys/fsext.h>
#include <libc/fsexthlp.h>
#include <sys/movedata.h>
#include <libc/farptrgs.h>
#include <libc/dosio.h>
#include <libc/getdinfo.h>
#include <libc/fd_props.h>
#include <go32.h>


static unsigned long _get_sft_entry_ptr(int fd)
{
  __dpmi_regs regs;
  unsigned char ind;
  unsigned long es, di;


  /* Get the JFT entry address for this handle.  */
  regs.x.ax = 0x1220;
  regs.x.bx = fd;
  __dpmi_int(0x2f, &regs);


  if (regs.x.flags & 1) /* int2F/1220 returns CF set if bad input */
  {
    errno = __doserr_to_errno(regs.h.al);  /* And AL has the error number */
    return 0;
  }


  /* Get the SFT entry number for this handle.  */
  es = regs.x.es;
  di = regs.x.di;
  ind = _farpeekb(_dos_ds, es * 16 + di);


  if (ind < 0xff)
  {
    /* Now get the address of the entry.  */
    regs.x.ax = 0x1216;
    regs.x.bx = ind;
    __dpmi_int (0x2f, &regs);
  }
  else
    regs.x.flags |= 1;  /* Set carry flag to simulate error       */

  if (regs.x.flags & 1) /* int2F/1216 returns CF set if bad input */
  {
    errno = EBADF;      /* But no other information is returned   */
    return 0;
  }


  es = regs.x.es;
  di = regs.x.di;


  return es * 16 + di;
}


static int
inherit_bit_test (int fd, short dev_info)
{
  __dpmi_regs regs;
  short new_dev_info;


  dev_info |= _DEV_NO_INHERIT;
  regs.x.ax = 0x4401;
  regs.x.bx = fd;
  regs.x.dx = dev_info;
  __dpmi_int(0x21, &regs);

  new_dev_info = _get_dev_info(fd);

  /* If the dev info words are equal, then the documented
     interface can be used to set the inheritance bit.  */
  return (new_dev_info == dev_info) ? 1 : 0;
}

static int
_get_SHARE_status (void)
{
  __dpmi_regs regs;

  regs.x.ax = 0x1000;
  __dpmi_int(0x2f, &regs);

  /* If al == 0xff then SHARE is installed, otherwise not */
  return (regs.h.al == 0xff) ? 1 : 0;
}

static int
_fcntl_lk64(int fd, int cmd, struct flock64 *lock_r64)
{
  int ret = -1;
  long long int ret64 = -1L;
  offset_t pos, cur_pos, lock_pos, len;
  long long int flen;

  /* Is this a directory? If so, fail. */
  if (__get_fd_flags(fd) & FILE_DESC_DIRECTORY)
  {
    errno = EINVAL;
    return -1;
  }

  /* Check if SHARE is loaded */
  ret = _get_SHARE_status();

  if (!ret) /* Then SHARE is NOT loaded, just return success */
  {
    if (cmd == F_GETLK64) /* Then make believe the lock is available */
      lock_r64->l_type = F_UNLCK;

    return ret;
  }
  else ret = -1; /* Restore default value */

  cur_pos = llseek(fd, 0L, SEEK_CUR);
  if (cur_pos < 0)
    return -1;      /* Assumes llseek has set errno */

  lock_pos = llseek (fd, lock_r64->l_start, lock_r64->l_whence);
  if (lock_pos < 0)
    return -1;      /* Assumes llseek has set errno */

  len = lock_r64->l_len;


  /* If l_len is zero, then the lock is to be set from the position
     represented by l_start/l_whence to the current end-of-file.
  */
  if (len == 0L)
  {
    flen = lfilelength(fd); /* Hold lfilelength in case len < 0 */
    len = flen - lock_pos;  /* len > 0      lock_pos before EOF
                               len == 0     lock_pos at     EOF
                               len < 0      lock_pos after  EOF
                            */

    /* If now len < 0, then lock_pos is beyond EOF, and
       the code below will calculate the correct region.
    */
    if (len < 0L)
    {
      pos = flen;   /* Start at EOF == lfilelength(fd) */
      lock_pos = flen; /* Also set lock_pos for later  */
      len = -len;   /* Make the length positive        */
    }
  }


  /* Return to saved position */
  ret64 = llseek (fd, cur_pos, SEEK_SET);
  if (ret64 < 0L)
    return -1;      /* Assumes llseek has set errno */
  else
    ret64 = -1;     /* Restore default value */


  /* If l_len is positive, the area to lock is from l_start
     to l_start + l_len - 1. If l_len is negative, the area to lock is
     from l_start + len to l_start - 1.

     Note that if l_len == 0, then pos and len may have already
     been set above, and lock_pos may have been reset to flen.
  */
  if (len > 0L)
  {
    pos = lock_pos;
  }
  else
  {
    pos = lock_pos + len;
    len = -len;
  }


  /* DOS/Win9x only support write locks via int 21/5C, so
     all read lock requests are treated like write locks
  */

  if (lock_r64->l_type == F_UNLCK)
  {
    ret = _dos_unlk64(fd, pos, len);
    if (ret != 0)
    {
      _doserrno = ret;
      errno = __doserr_to_errno(ret);
      return -1;
    }
  }
  else if ((lock_r64->l_type == F_WRLCK) ||
           (lock_r64->l_type == F_RDLCK))
  {
    ret = _dos_lk64(fd, pos, len);
    if (cmd == F_GETLK64)
    {
      if (!ret)
      {
        _dos_unlk64(fd, pos, len);
        /* If no lock is found that would prevent a lock from
           being created, the lock type is set to F_UNLCK.
        */
        lock_r64->l_type = F_UNLCK;
      }
      else
      {
        /* If a lock is found then l_whence, l_start, and l_len
           should point to the area covered by the lock. But the
           file locking interface doesn't report where the
           existing lock is, so nothing more can be done.
        */
        _doserrno = ret;
        errno = __doserr_to_errno(ret);
        return -1;
      }
    }
    else
    {
      /* If F_SETLKW64 is set, wait for the lock to be released.  */
      if (cmd == F_SETLKW64 && ret)
      {
        while ((ret = _dos_lk64(fd, pos, len)))
          __dpmi_yield();
      }
      else if (ret)
      {
        _doserrno = ret;
        errno = __doserr_to_errno(ret);
        return -1;
      }
    }
  }
  if (ret)
    errno = ENOSYS;

  return ret;
}


int
fcntl(int fd, int cmd, ...)
{
  int tofd, open_max;
  va_list ap;
  __FSEXT_Function *func;
  short dev_info = _get_dev_info(fd);
  static int inherit_bit_visible = -1;
  int errno_save;

  /* Verify the descriptor is valid by retrieving
     the handle's device info word.  */
  if (dev_info == -1)
    return dev_info;


  /* Allow a fd to override with a FSEXT.  */
  func = __FSEXT_get_function(fd);
  if (func)
  {
    int rv;
    if (__FSEXT_func_wrapper(func, __FSEXT_fcntl, &rv, fd))
      return rv;
  }


  errno_save = errno;

  switch (cmd)
  {
    case F_DUPFD:
    {
      va_start(ap, cmd);
      tofd = va_arg(ap, int);
      va_end(ap);


      open_max = getdtablesize();
      while (tofd < open_max)
      {
        /* If unable to get the device info for the handle,
           then the handle is not active and it can be used.  */
        if (_get_dev_info(tofd) == -1)
          break;
        tofd++;
      }


      if (tofd >= open_max)
      {
        errno = EMFILE;
        return -1;
      }


      errno = errno_save;
      return dup2(fd, tofd);
    }


    case F_GETFD:
    {
       unsigned long entry_ptr;


      /* DOS only passes the first 20 handles to child programs.  In
         addition, handles 19 and 18 will be closed by the stub of the
         child program (if it is a DJGPP program).  */


      if (fd >= 18)
        return FD_CLOEXEC;


      /* Determine if the documented interface will allow twiddling with
         the inherit bit. If not, fallback to the undocumented one.  */
      if (inherit_bit_visible == -1)
        inherit_bit_visible = inherit_bit_test (fd, dev_info);

      if (!inherit_bit_visible)
      {
        entry_ptr = _get_sft_entry_ptr(fd);
        if (entry_ptr == 0)
        {
          /* The fd has already been validated, so reaching here means
             something is wrong with _get_sft_entry_ptr. */
          return -1;
        }
        /* Offset 5 in the SFT contains the device info word.  */
        dev_info = _farpeekw(_dos_ds, entry_ptr + 5);
      }
      return (dev_info & _DEV_NO_INHERIT) ? FD_CLOEXEC : 0;
    }


    case F_SETFD:
    {
      unsigned int flag;
      unsigned long entry_ptr = 0; /* shut up -Wall */
      __dpmi_regs regs;


      va_start (ap, cmd);
      flag = va_arg(ap, int);
      va_end(ap);


      /* DOS only passes the first 20 handles to child programs.  In
         addition, handles 19 and 18 will be closed by the stub of the
         child program (if it is a DJGPP program).  */
      if (fd >= 18)
      {
        if (flag & FD_CLOEXEC)
          return 0;
        else
        {
          errno = ENOSYS;
          return -1;
        }
      }


      /* Determine if the documented interface will allow twiddling with
         the inherit bit. If not, fallback to the undocumented one.  */
      if (inherit_bit_visible == -1)
        inherit_bit_visible = inherit_bit_test(fd, dev_info);

      if (!inherit_bit_visible)
      {
        entry_ptr = _get_sft_entry_ptr(fd);
        if (entry_ptr == 0)
        {
          /* The fd has already been validated, so reaching here means
             something is wrong with _get_sft_entry_ptr. */
          return -1;
        }


        dev_info = _farpeekw(_dos_ds, entry_ptr + 5);
      }


      if (flag & FD_CLOEXEC)
        dev_info |= _DEV_NO_INHERIT;
      else
        dev_info &= ~(_DEV_NO_INHERIT);


      if (inherit_bit_visible)
      {
        regs.x.ax = 0x4401;
        regs.x.bx = fd;
        regs.x.dx = dev_info;
        __dpmi_int(0x21, &regs);
      }
      else
        _farpokew(_dos_ds, entry_ptr + 5, dev_info);


      return 0;
    }


    case F_GETFL:
    {
      unsigned long sft_entry_ptr;
      unsigned char sft_open_flags;
      int flags = 0;

      /*
       * Use the data in the SFT.
       *
       * This seems to work on Windows 2000 and therefore probably works
       * on Windows XP. It does not work on Windows NT 4, where the file
       * will always appear to have been opened read-only. It's hard
       * to distinguish Windows 2000 and Windows NT 4 with an LFN TSR,
       * so we can't have a special case for Windows NT 4.
       */
      sft_entry_ptr = _get_sft_entry_ptr(fd);
      if (sft_entry_ptr)
      {
	/* Offset 2 in the SFT contains the open flags. */
	sft_open_flags = _farpeekw(_dos_ds, sft_entry_ptr + 2);

	switch(sft_open_flags & 0x7)
	{
	case 0: flags |= O_RDONLY; break;
	case 1: flags |= O_WRONLY; break;
	case 2: flags |= O_RDWR;   break;

	default:
	  break;
	}
      }
      else
      {
	/* If we've failed to get an SFT pointer, pretend that the file
	 * is read-write, since that's most likely to be correct. */
	flags |= O_RDWR;
      }

      /* Check the FILE_DESC_APPEND flag in __fd_properties. */
      if (__has_fd_properties(fd))
      {
	unsigned long fd_flags = __get_fd_flags(fd);

	if (fd_flags & FILE_DESC_APPEND)
	  flags |= O_APPEND;
      }

      return flags;
    }


    case F_SETFL:
    {
      unsigned char new_mode_bits;


      va_start (ap, cmd);
      new_mode_bits = va_arg(ap,int);
      va_end(ap);


      /* Allow removal of O_NONBLOCK, since DJGPP doesn't support it
         anyway.  */
      return (new_mode_bits == 0) ? 0 : -1;
    }


    case F_GETLK:
    case F_SETLK:
    case F_SETLKW:
    {
      struct flock *lock_req = NULL; /* shut up -Wall */
      struct flock64 lock_r64;
      int ret = -1, lcmd = -1;

      va_start (ap, cmd);
      lock_req = va_arg(ap, struct flock *);
      va_end (ap);

      switch (cmd)
      {
        case F_GETLK:   lcmd = F_GETLK64;  break;
        case F_SETLK:   lcmd = F_SETLK64;  break;
        case F_SETLKW:  lcmd = F_SETLKW64; break;
      }

      lock_r64.l_type   = lock_req->l_type;
      lock_r64.l_start  = lock_req->l_start;
      lock_r64.l_whence = lock_req->l_whence;
      lock_r64.l_len    = lock_req->l_len;

      ret = _fcntl_lk64(fd, lcmd, &lock_r64);

      lock_req->l_type   =         lock_r64.l_type;
      lock_req->l_start  = (off_t) lock_r64.l_start;
      lock_req->l_whence =         lock_r64.l_whence;
      lock_req->l_len    = (off_t) lock_r64.l_len;

      return ret;
    }


    case F_GETLK64:
    case F_SETLK64:
    case F_SETLKW64:
    {
      struct flock64 *lock_r64; /* shut up -Wall */
      int ret;

      va_start (ap, cmd);
      lock_r64 = va_arg(ap, struct flock64 *);
      va_end (ap);

      ret = _fcntl_lk64(fd, cmd, lock_r64);

      return ret;
    }
  }


  /* In case fcntl is called with an unrecognized command.  */
  errno = ENOSYS;
  return -1;
}
