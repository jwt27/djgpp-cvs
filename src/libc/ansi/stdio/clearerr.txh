@node clearerr, stdio
@findex clearerr
@subheading Syntax

@example
#include <stdio.h>

void clearerr(FILE *stream);
@end example

@subheading Description

This function clears the EOF and error indicators for the file
@var{stream}. 

@subheading Return Value

None.

@subheading Portability

@portability ansi, posix

@subheading Example

@example
clearerr(stdout);
@end example

