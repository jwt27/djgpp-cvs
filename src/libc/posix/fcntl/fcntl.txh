@node fcntl, io
@findex fcntl
@subheading Syntax

@example
#include <fcntl.h>

int fcntl (int fd, int cmd, ...);
@end example

@subheading Description

This function performs the operation specified by @var{cmd} on
the file open on handle @var{fd}.  The following operations are defined
by the header @code{fcntl.h}:

@table @code
@item F_DUPFD
Returns a file handle that duplicates @var{fd} like @code{dup} does
(@pxref{dup}), except that @code{fcntl} also makes sure the returned
handle is the lowest available handle greater than or equal to the
integer value of the third argument.

@item F_GETFD
Get the @code{FD_CLOEXEC} close-on-exec (a.k.a.@: no-inherit) status of
@var{fd}.  If the returned value has its least-significant bit set, the
file will not be inherited by programs invoked by this process;
otherwise, the file will remain open in the child processes.

Note that only the first 20 handles can be passed to child processes by
DOS/Windows; handles beyond that cannot be inherited.  In addition, the
stub loader of the child DJGPP program will forcibly close handles 19
and 18 (since otherwise it will be unable to read the COFF executable
information and enter protected mode).  Therefore, the current
implementation always returns @code{FD_CLOEXEC} for handles 18 and
above.

For handles less than 18, the call will try to determine the status of
the @code{O_NOINHERIT} flag for that file and will return either
@code{FD_CLOEXEC} if the flag is set, or 0 if the flag is not set.  If
the status of the @code{O_NOINHERIT} flag cannot be determined, the call
will return -1, setting @code{errno} to @code{ENOSYS}.

The no-inherit bit can be set when the file is opened by using the
@code{O_NOINHERIT} in the open flags; see @ref{open}.

@item F_SETFD
Set or unset the close-on-exec flag for the handle @var{fd} using the
LSB of the integer value supplied as the third argument.  Since only the
first 20 handles are passed to child programs, and since the stub loader
of the child DJGPP program will forcibly close handles 19 and 18 (since
otherwise it will be unable to read the COFF executable information and
enter protected mode), the flag can only be set or unset on the first 18
handles.  Attempts to set the flag for handles 18 or above will always
return 0, and attempts to unset the flag for handles 18 or above will
always return -1, setting @code{errno} to @code{ENOSYS}.

For handles less than 18, the call will try to set or unset the
@code{O_NOINHERIT} flag for that file and will return 0 if the flag is
changed.  If the @code{O_NOINHERIT} flag cannot be changed, the call
will return -1, setting @code{errno} to @code{ENOSYS}.
  
@item F_GETFL
Get the open mode and status flags associated with the handle @var{fd}.
The flags are those supported by @code{open} and @code{creat} functions,
like @code{O_RDONLY}, @code{O_APPEND}, etc.

On Windows NT this cannot report the open mode correctly ---
@code{O_RDONLY} is always returned.

@item F_SETFL
Set the open mode and status flags associated with the handle @var{fd}.
This fails in all but one case, and sets @code{errno} to @code{ENOSYS},
since DOS and Windows don't allow changing the descriptor flags after
the file is open.

The one allowed case is for @code{O_NONBLOCK}, since DJGPP doesn't
support it anyway.  That is, calls using @code{F_SETFL} will fail for
all flag values @strong{except} @code{O_NONBLOCK}.

@example
#include <fcntl.h>

ret = fcntl(fd, F_SETFL, O_BINARY); /* This will fail, returning -1  */
                                    /* and setting errno to ENOSYS.  */

ret = fcntl(fd, F_SETFL, O_NONBLOCK); /* This will succeed          */
                                      /* returning 0.               */
@end example

@item F_GETLK
Return the lock structure that prevents obtaining the lock pointed to by
the third argument, or set the @code{l_type} field of the lock structure
to @code{F_UNLCK} if there is no obstruction.  Currently, only the
setting of the @code{l_type} field is provided.  This call will not
return values in the @code{struct flock} parameter identifying what lock
parameters prevent getting the requested lock, since there is no way to
obtain this information from DOS/Windows.  If the lock cannot be
obtained, -1 is returned and @code{errno} is set to the reason (which
will be one of @code{EINVAL}, @code{EBADF}, @code{EACCES} or
@code{ENOLCK}).

Locking of directories is not supported.

@item F_SETLK
Set or clear a file segment lock according to the structure pointed to
by the third argument.  The lock is set when @code{l_type} is
@code{F_RDLCK} (shared lock request) or @code{F_WRLCK} (exclusive lock
request), and the lock is cleared when @code{l_type} is @code{F_UNLCK}.
If the lock is already held, then this call returns -1 and sets
@code{errno} to @code{EACCES}.

The @code{F_RDLCK} value for requesting a read lock is always treated as
if it were @code{F_WRLCK} for a write lock.

This is because DOS/Win9x only supports one kind of lock, and it is the
exclusive kind.

Locking of directories is not supported.

@item F_SETLKW
Same as @code{F_SETLK}, but if the lock is blocked, the call will wait
(using @code{__dpmi_yield}, see @pxref{__dpmi_yield}) until it is
unblocked and the lock can be applied.  This call will @strong{never
exit} if the program making the call is the program which already owns
the lock.

Locking of directories is not supported.

@item F_GETLK64
@item F_SETLK64
@item F_SETLKW64
Each of these does exactly the same function as the non-"64" version,
but the third argument must be of type @code{struct flock64}, which
allows the @var{l_start} and @var{l_len} members to be @code{long long
int} values.  The current code will only use these @code{long long int}
values modulo @math{2^32}, which allows file locking positions up to 4
gigabytes minus 1.  True 64-bit file locking is @strong{not} supported.

The @code{struct flock64} members @var{l_start} and @var{l_len} are
declared to be of type @code{offset_t}, which is in turn typedef'ed to
be a @code{long long}.

Locking of directories is not supported.
@end table

This function can be hooked by the @dfn{Filesystem extensions}, see
@ref{File System Extensions}.  If you don't want this, and you are
calling @code{fcntl} with the @code{F_DUPFD} command, you should use
@code{dup2} instead, see @ref{dup2}.

@subheading Return Value

If an invalid or unsupported value is passed in @var{cmd}, or @var{fd}
is an invalid file handle, the function returns -1 and sets @code{errno}
to the appropriate value.  Unsupported values of @var{cmd} cause
@code{ENOSYS} to be stored in @code{errno}.  If @var{cmd} is
@code{F_DUPFD}, the function returns the new descriptor or -1 in case of
a failure.

Lock requests which specify the open file's current @code{EOF} position
as the value of @var{l_start} and zero as the @var{l_len} value will
fail, returning -1 with @code{errno} set to @code{EACCES}.


@subheading Portability

@port-note posix Contrary to Posix requirement, the handle returned by @code{F_DUPFD} shares the @code{FD_CLOEXEC} flag with @var{fd} (unless they are on different sides of the 20-handle mark), since DOS/Windows only maintain a single set of bits for all the handles associated with the same call to @code{open}.
@portability !ansi, posix


@subheading Example

@example
 /* Save the handle in a way that it won't be passed
    to child processes.  */
  int saved_fd = fcntl(fd, F_DUPFD, 20);

 /* Set an advisory lock for the whole file.  */
  struct flock flock;
  int retval, fd;

  flock.l_type = F_RDLCK;
  flock.l_whence = SEEK_SET;
  flock.l_start = flock.l_len = 0;
  errno = 0;
  retval = fcntl(fd, F_SETLK, &flock);

 /* Get the status of the lock we just obtained
    (should return -1 with errno == EACCES).  */
  errno = 0;
  retval = fcntl(fd, F_GETLK, &flock);

 /* Release the lock.  */
  errno = 0;
  flock.l_type = F_UNLCK;
  retval = fcntl(fd, F_SETLK, &flock);

 /* Get the status of the lock we just released
    (should return 0).  */
  errno = 0;
  flock.l_type = F_RDLCK;
  retval = fcntl(fd, F_GETLK, &flock);

 /* Try to set the O_BINARY flag on the open file
    (should return -1 with errno == ENOSYS).  */
  errno = 0;
  retval = fcntl(fd, F_SETFL, O_BINARY);

 /* Set the O_NONBLOCK flag on the open file
    (should return 0).  */
  errno = 0;
  retval = fcntl(fd, F_SETFL, O_NONBLOCK);

 /* Get the flags on the open file
    (always returns 0).  */
  errno = 0;
  retval = fcntl(fd, F_GETFL);

@end example
