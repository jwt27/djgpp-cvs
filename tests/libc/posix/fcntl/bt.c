#include <stdio.h>
#include <fcntl.h>
#ifdef __GO32__
#include <unistd.h>
#endif
#include <io.h>

#define BS 10
#define TMP "bt.dat"

int
mk(const char *s, int m)
{
  int fd = open(TMP, O_WRONLY | O_CREAT | O_TRUNC | m, 0600);
  int i;
  char buf[4000];
  int b=0;
  for (i=0; s[i]; i++)
    switch (s[i])
    {
    case 'r':
      buf[b++] = '\r';
      break;
    case 'n':
      buf[b++] = '\n';
      break;
    case 'z':
      buf[b++] = 26;
      break;
    case '0':
      buf[b++] = 0;
      break;
    default:
      buf[b++] = s[i];
      break;
    }
  b = write(fd, buf, b);
  printf("%d bytes written\n", b);
  close(fd);
  return 0;
}

void
p1(int c)
{
  switch (c)
  {
  case 0:
    printf(" \\0");
    break;
  case '\r':
    printf(" \\r");
    break;
  case 26:
    printf(" \\Z");
    break;
  case '\n':
    printf(" \\n");
    break;
  default:
    if (c < ' ')
      printf(" %02x", c & 0xff);
    else
      printf(" %c", c);
    break;
  }
}

void
binary_test(void)
{
  char buf[20];
  int i, r;
  int fd = open(TMP, O_RDONLY|O_BINARY);
  r = read(fd, buf, 20);
  printf("binary[%03d]:", r);
  for (i=0; i<r; i++)
    p1(buf[i]);
  printf("\n");
  close(fd);
}

void
read_test(void)
{
  char buf[BS];
  int i, r;
  int fd = open(TMP, O_RDONLY|O_TEXT);
  r = read(fd, buf, BS);
  printf("  read[%03d]:", r);
  for (i=0; i<r; i++)
    p1(buf[i]);
  printf("\n");
  close(fd);
}

void
fread_test(void)
{
  char buf[BS];
  int i, r;
  FILE *f = fopen(TMP, "rt");
  r = fread(buf, 1, BS, f);
  printf(" fread[%03d]:", r);
  for (i=0; i<r; i++)
    p1(buf[i]);
  printf("\n");
  fclose(f);
}

void
fgetc_test(void)
{
  int i;
  FILE *f = fopen(TMP, "rt");
  printf("      fgetc:");
  while ((i=fgetc(f)) != EOF)
    p1(i);
  printf("\n");
  fclose(f);
}

int
main(int argc, char **argv)
{
  const char *ptr = "xxxrnrnnrxnxnrnxrnrrrnxnnxrnrnrnxrnrxnxnr";
  if (argc > 1)
    ptr = argv[1];

  printf("Request Size = %d\n", BS);

  printf("Binary write...\n");
  mk(ptr, O_BINARY);
  binary_test();
  read_test();
  fread_test();
  fgetc_test();

  printf("Text write...\n");
  mk(ptr, O_TEXT);
  binary_test();
  read_test();
  fread_test();
  fgetc_test();

  remove(TMP);
  return 0;
}
