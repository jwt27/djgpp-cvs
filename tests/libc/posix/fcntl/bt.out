Request Size = 10
Binary write...
41 bytes written
binary[020]: x x x \r \n \r \n \n \r x \n x \n \r \n x \r \n \r \r
  read[007]: x x x \n \n \n x
 fread[010]: x x x \n \n \n x \n x \n
      fgetc: x x x \n \n \n x \n x \n \n x \n \n x \n \n x \n \n \n x \n x \n x \n
Text write...
41 bytes written
binary[020]: x x x \r \r \n \r \r \n \r \n \r x \r \n x \r \n \r \r
  read[005]: x x x \n \n
 fread[010]: x x x \n \n \n x \n x \n
      fgetc: x x x \n \n \n x \n x \n \n x \n \n x \n \n x \n \n \n x \n x \n x \n
